storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java

```diff
@@ -61,29 +62,96 @@ public JdbcStorageManager(QueryExecutor queryExecutor) {
 
     @Override
     public void add(Storable storable) throws AlreadyExistsException {
-        log.debug("Adding storable [{}]", storable);
-        queryExecutor.insert(storable);
+        boolean committed = false;
+        try {
+            beginTransaction(TransactionIsolation.DEFAULT);
+
+            log.debug("Adding storable [{}]", storable);
+            final Storable existing = get(storable.getStorableKey());
+            if (existing == null) {
+                queryExecutor.insert(storable);
+            } else if (!existing.equals(storable)) {
+                throw new AlreadyExistsException("Another instance with same id = " + storable.getPrimaryKey()
+                        + " exists with different value in namespace " + storable.getNameSpace()
+                        + " Consider using addOrUpdate method if you always want to overwrite.");
+            }
+
+            commitTransaction();
+            committed = true;
+        } finally {
+            if (!committed) {
+                rollbackTransaction();
+            }
+        }
     }
 
     @Override
     public <T extends Storable> T remove(StorableKey key) throws StorageException {
-        T oldVal = get(key);
-        if (key != null) {
-            log.debug("Removing storable key [{}]", key);
-            queryExecutor.delete(key);
+        boolean committed = false;
+        try {
+            beginTransaction(TransactionIsolation.DEFAULT);
+
+            T oldVal = get(key);
```

Comment:
get() here may not be required. As by default "delete" command with "where" clause which doesn't match any records when executed doesn't throw any error, but quietly prints that "0 rows deleted". I confirmed this behavior in MySQL, Postgres and Oracle.

Reply:
The method needs to return old value as method contract / signature, so need to get old value first.

Reply:
I see that there are instances were storage entity obtained by calling remove() is being used and certain places where we just discard the value. I think we should have remove() not return any values and if the user wants to fetch the entity which is being removed then he can do get() and followed by remove(). I know this is an existing behavior not related to this PR, if you agree with the proposal we can come back to it in a different PR.